src/Container.ts

```diff
@@ -96,18 +106,29 @@ export class Container {
         if (!type && name)
             throw new Error(`Service named ${name} was not found, probably it was not registered`);
 
-        // if params are given we need to go throw each of them and initialize them all properly
+        let objectInstance: any;
+
+        const factory = this.findRegisteredFactoryByType(type);
+
+        if (factory && factory.params) {
+            params = factory.params;
+        }
+
+        // if params are given we need to go through each of them and initialize them all properly
         if (params) {
             params = this.initializeParams(type, params);
-            params.unshift(null);
+            if (!factory) {
+                params.unshift(null);
+            }
         }
 
-        let objectInstance: any;
-
         if (factory) {
-            // Calling factory function if specified to create an instance.
-            objectInstance = factory.apply(undefined, params);
+
+            // Calling factory method to create an instance.
+            objectInstance = factory.factoryFunction.apply(undefined, params);
```

Comment:
it can be `factory.factoryFunction(...params)` instead. Btw is there a better name for `factory.factoryFunction` ?